packages/core/elasticsearch/core-elasticsearch-client-server-internal/src/agent_manager.ts

```diff
@@ -0,0 +1,188 @@
+  /**
+   * Store Agent instances by type and protocol, e.g.:
+   *
+   * data: {
+   *   'http:': [agentInstance1, agentInstance2],
+   *   'https:': [agentInstance3, agentInstance4]
+   * },
+   * monitoring: {
+   *   'http:': [],
+   *   'https:': [agentInstance5]
+   * }
+   */
+  private agentStore: Record<string, Record<Protocol, Array<NetworkAgent | undefined>>>;
+  private logger: Logger;
+
+  constructor(
+    logger: Logger,
+    private defaultConfig: HttpAgentOptions = {
+      keepAlive: true,
+      keepAliveMsecs: 50000,
+      maxSockets: 256,
+      maxFreeSockets: 256,
+      scheduling: 'lifo',
+    }
+  ) {
+    this.logger = logger;
+    this.agentStore = {};
+  }
+
+  public getAgentFactory(
+    type: string,
+    agentOptions?: HttpAgentOptions | UndiciAgentOptions | AgentFactory | false
+  ): AgentFactory {
+    if (isAgentFactory(agentOptions)) {
+      // use the user-provided factory directly
+      return agentOptions;
+    }
+
+    const agentConfig = assertValidAgentConfig(agentOptions);
+
+    // a given agent factory (of a given type) always provides the same Agent instances (for the same protocol)
+    // we keep the indexes for each protocol, so that we can access them later on, when the factory is invoked
+    const { httpInstanceIndex, httpsInstanceIndex } = this.initTypeStore(type);
+
+    return (connectionOpts: ConnectionOptions): NetworkAgent => {
+      const agents = this.getAgentInstances(type, connectionOpts);
+
+      // when the factory is called for a given connection (and protocol), we use the appropriate index
+      const agentIndex = isHttps(connectionOpts) ? httpsInstanceIndex : httpInstanceIndex;
+
+      let agent = agents[agentIndex];
+
+      if (!agent) {
+        agent = this.createAgent(agentConfig, connectionOpts);
+
+        // Allow GC'ing the Agent after it has been terminated
+        const doDestroy = agent.destroy.bind(agent);
+        agent.destroy = () => {
+          doDestroy();
+          agents[agentIndex] = undefined;
+        };
+
+        // add the new Agent instance to the list of instances of that type
+        agents[agentIndex] = agent;
+      }
+
+      return agent;
+    };
+  }
+
+  private initTypeStore(type: string): { httpInstanceIndex: number; httpsInstanceIndex: number } {
+    let agentsOfType = this.agentStore[type];
+
+    if (!agentsOfType) {
+      agentsOfType = {
+        [HTTP]: [],
+        [HTTPS]: [],
+      };
+
+      this.agentStore[type] = agentsOfType;
+    } else {
+      this.logger.warn(
+        `An Agent factory for type ${type} has already been retrieved.
+        This is not optimal since the existing instance cannot be reused.
+        Please consider reusing the previously retrieved factory instead`
+      );
+    }
+
+    const httpInstanceIndex = agentsOfType[HTTP].length;
+    const httpsInstanceIndex = agentsOfType[HTTPS].length;
+
+    agentsOfType[HTTP].push(undefined);
+    agentsOfType[HTTPS].push(undefined);
```

Comment:
I understand the problem at stake here, but I'm not a big fan of pushing indefined's into a supposedly typed array here. 

I think I would have used a map with a `{type}-{protocol}-{index}`, and/or kept the current index (indices?) per type stored somewhere to avoid that.

But this is an personal implementation preference here, what you did works (as long as we don't exceed `MAX_SAFE_INTEGER` agents instantiated for a given type)

Reply:
Fair enough, pushing `undefined` just to book a slot in the isn't the nicest code ever, I'll give you that 😛 .

Reply:
I realised that type was of no use to the AgentManager ATM, so I updated the PR with a much simpler version of the Agent, which only has a couple of arrays now:

* one for Http agent instances
* one for Https agent instances